core/src/mast/multi_forest_node_iterator.rs

```diff
@@ -0,0 +1,395 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_crypto::hash::rpo::RpoDigest;
+
+use crate::mast::{MastForest, MastForestError, MastNode, MastNodeId};
+
+type ForestIndex = usize;
+
+/// Depth First Search Iterator in Post Order for [`MastForest`]s.
+///
+/// This iterator iterates through all **reachable** nodes of a forest exactly once.
+///
+/// Since a `MastForest` has multiple possible entrypoints in the form of its roots, a depth-first
+/// search must visit all of those roots and the trees they form.
+///
+/// For instance, consider this `MastForest`:
+///
+/// ```text
+/// Nodes: [Block(foo), Block(bar), Join(0, 1), External(qux)]
+/// Roots: [2]
+/// ```
+///
+/// The only root is the `Join` node at index 2. The first three nodes of the forest form a
+/// tree, since the `Join` node references index 0 and 1. This tree is discovered by
+/// starting at the root at index 2 and following all children until we reach terminal nodes (like
+/// `Block`s) and build up a stack of the discovered, but unvisited nodes. The stack is
+/// built such that popping elements off the stack (from the back) yields a postorder.
+///
+/// After the first tree is discovered, the stack looks like this: `[2, 1, 0]`. On each
+/// call to `next` one element is popped off this stack and returned.
+///
+/// If the stack is exhausted we start another discovery if more unvisited roots exist. Since the
+/// `External` node is not a root and not referenced by any other tree in the forest, it will not be
+/// visited.
+///
+/// The iteration on a high-level thus consists of a constant back and forth between discovering
+/// trees and returning nodes from the stack.
+///
+/// Note: This type could be made more general to implement pre-order or in-order iteration too.
+pub(crate) struct MultiMastForestNodeIter<'forest> {
+    /// The forest that we're iterating.
+    mast_forests: Vec<&'forest MastForest>,
+    /// The procedure root index at which we last started a tree discovery.
+    ///
+    /// This value iterates through 0..mast_forest.num_procedures() which guarantees that we visit
+    /// all nodes reachable from all roots.
+    last_forest_idx: usize,
+    last_procedure_root_idx: u32,
+    non_external_nodes: BTreeMap<RpoDigest, (usize, MastNodeId)>,
+    /// Describes whether the node at some index has already been visited. Note that this is set to
+    /// true for all nodes on the stack, even if the caller of the iterator has not yet seen the
+    /// node. See [`Self::visit_later`] for more details.
+    node_visited: Vec<Vec<bool>>,
+    /// This stack always contains the discovered but unvisited nodes.
+    /// For any id store on the stack it holds that `node_visited[id] = true`.
+    unvisited_node_stack: Vec<MultiMastForestIteratorItem>,
+}
+
+impl<'forest> MultiMastForestNodeIter<'forest> {
+    pub(crate) fn new(mast_forests: Vec<&'forest MastForest>) -> Self {
```

Comment:
Could we add brief doc-comments to this constructor? My understanding is that here we make the first pass over all nodes and build a map of all non-external nodes, right?

Reply:
This is where we build a map of all non-external nodes, so when we encounter an External node with digest `x` we can easily check if any forest has a node with digest `x`!
I'm in the process of adding documentation everywhere now :+1: since I didn't get around to it yesterday.